metagraph/src/annotation/int_matrix/row_diff/int_row_diff.hpp

```diff
@@ -0,0 +1,343 @@
+
+    const graph::boss::BOSS &boss = graph_->get_boss();
+    const bit_vector &rd_succ = fork_succ_.size() ? fork_succ_ : boss.get_last();
+
+    for (size_t i = 0; i < row_ids.size(); ++i) {
+        Row row = row_ids[i];
+
+        graph::boss::BOSS::edge_index boss_edge = graph_->kmer_to_boss_index(
+                graph::AnnotatedSequenceGraph::anno_to_graph_index(row));
+
+        while (true) {
+            row = graph::AnnotatedSequenceGraph::graph_to_anno_index(
+                    graph_->boss_to_kmer_index(boss_edge));
+
+            auto [it, is_new] = node_to_rd.try_emplace(row, rd_ids.size());
+            rd_paths_trunc[i].push_back(it.value());
+
+            // If a node had been reached before, we interrupt the diff path.
+            // The annotation for that node will have been reconstructed earlier
+            // than for other nodes in this path as well. Thus, we will start
+            // reconstruction from that node and don't need its successors.
+            if (!is_new)
+                break;
+
+            rd_ids.push_back(row);
+
+            if (anchor_[row])
+                break;
+
+            boss_edge = boss.row_diff_successor(boss_edge, rd_succ);
+        }
+    }
+
+    node_to_rd = VectorMap<Row, size_t>();
+
+    std::vector<RowValues> rd_rows = diffs_.get_row_values(rd_ids);
+    for (auto &row : rd_rows) {
+        decode_diffs(&row);
+    }
+
+    rd_ids = std::vector<Row>();
+
+    // reconstruct annotation rows from row-diff
+    std::vector<RowValues> rows(row_ids.size());
+
+    for (size_t i = 0; i < row_ids.size(); ++i) {
+        RowValues &result = rows[i];
+        // propagate back and reconstruct full annotations for predecessors
+        for (auto it = rd_paths_trunc[i].rbegin(); it != rd_paths_trunc[i].rend(); ++it) {
+            std::sort(rd_rows[*it].begin(), rd_rows[*it].end());
+            add_diff(rd_rows[*it], &result);
+            // replace diff row with full reconstructed annotation
+            rd_rows[*it] = result;
+        }
+        assert(std::all_of(result.begin(), result.end(),
+                           [](auto &p) { return p.second; }));
+        assert(std::all_of(result.begin(), result.end(),
+                           [](auto &p) { return (int64_t)p.second > 0; }));
+    }
+
+    return rows;
+}
+
+template <class BaseMatrix>
+bool IntRowDiff<BaseMatrix>::load(std::istream &in) {
+    std::string version(4, '\0');
+    in.read(version.data(), 4);
+    return anchor_.load(in) && fork_succ_.load(in) && diffs_.load(in);
+}
+
+template <class BaseMatrix>
+void IntRowDiff<BaseMatrix>::serialize(std::ostream &out) const {
+    out.write("v2.0", 4);
+    anchor_.serialize(out);
+    fork_succ_.serialize(out);
+    diffs_.serialize(out);
+}
+
+template <class BaseMatrix>
+void IntRowDiff<BaseMatrix>::decode_diffs(RowValues *diffs) {
+    for (auto &[j, value] : *diffs) {
+        value = decode_diff(value);
```

Comment:
maybe make it more clear that this is doing a signed->unsigned conversion
```suggestion
        value = static_cast<uint64_t>(decode_diff(value));
```

Reply:
It's done automatically

Reply:
I know, I was only suggesting this to make it more clear to the reader